internal/campaigns/run_steps.go

```diff
@@ -87,29 +94,74 @@ func runSteps(ctx context.Context, wc *WorkspaceCreator, repo *graphql.Repositor
 
 		// Set up a temporary file on the host filesystem to contain the
 		// script.
-		fp, err := ioutil.TempFile(tempDir, "")
+		runScriptFile, err := ioutil.TempFile(tempDir, "")
 		if err != nil {
 			return nil, errors.Wrap(err, "creating temporary file")
 		}
-		hostTemp := fp.Name()
-		defer os.Remove(hostTemp)
-		if _, err := fp.WriteString(step.Run); err != nil {
-			return nil, errors.Wrapf(err, "writing to temporary file %q", hostTemp)
+		defer os.Remove(runScriptFile.Name())
+
+		// Parse step.Run as a template...
+		tmpl, err := parseAsTemplate("step-run", step.Run, &stepContext)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing step run")
+		}
+
+		// ... and render it into a buffer and the temp file we just created.
+		var runScript bytes.Buffer
+		if err := tmpl.Execute(io.MultiWriter(&runScript, runScriptFile), stepContext); err != nil {
+			return nil, errors.Wrap(err, "executing template")
+		}
+		runScriptFile.Close()
+
+		// Parse and render the step.Files.
+		files, err := renderStepFiles(step.Files, &stepContext)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing step files")
+		}
+
+		// Create temp files with the rendered content of step.Files so that we
+		// can mount them into the container.
+		filesToMount := make(map[string]*os.File, len(files))
+		for name, content := range files {
+			fp, err := ioutil.TempFile(tempDir, "")
+			if err != nil {
+				return nil, errors.Wrap(err, "creating temporary file")
+			}
+			if _, err := io.Copy(fp, content); err != nil {
+				return nil, errors.Wrap(err, "writing to temporary file")
+			}
+
+			filesToMount[name] = fp
+		}
+
+		// Render the step.Env variables as templates.
+		env, err := renderStepEnv(step.Env, &stepContext)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing step files")
 		}
-		fp.Close()
 
+		reportProgress(runScript.String())
 		const workDir = "/work"
-		cmd := exec.CommandContext(ctx, "docker", "run",
+		args := []string{
+			"run",
 			"--rm",
 			"--cidfile", cidFile.Name(),
 			"--workdir", workDir,
 			"--mount", fmt.Sprintf("type=bind,source=%s,target=%s", volumeDir, workDir),
-			"--mount", fmt.Sprintf("type=bind,source=%s,target=%s,ro", hostTemp, containerTemp),
-			"--entrypoint", shell,
-		)
-		for k, v := range step.Env {
-			cmd.Args = append(cmd.Args, "-e", k+"="+v)
+			"--mount", fmt.Sprintf("type=bind,source=%s,target=%s,ro", runScriptFile.Name(), containerTemp),
 		}
+		for target, source := range filesToMount {
+			args = append(args, "--mount", fmt.Sprintf("type=bind,source=%s,target=%s,ro", source.Name(), target))
+			defer os.Remove(source.Name())
```

Comment:
Is there a reason to not have this `defer` immediately after creating the temporary file above?

Reply:
Yes: I forgot. I kinda reorganised that code a bit and I think I left this lying around here. Good catch! Fixed.